modules/navier_stokes/include/userobjects/INSFVRhieChowInterpolator.h

```diff
@@ -0,0 +1,275 @@
+  const std::set<SubdomainID> _sub_ids;
+
+  /// A map from element IDs to 'a' coefficient data
+  std::unordered_map<dof_id_type, libMesh::VectorValue<ADReal>> _a;
+
+  /// A map from element IDs to 'B' data. After populating the map, this object can be evaluated at
+  /// both element and face centers
+  CellCenteredMapFunctor<libMesh::VectorValue<ADReal>,
+                         std::unordered_map<dof_id_type, libMesh::VectorValue<ADReal>>>
+      _b;
+
+  /// The result of one interpolation and reconstruction of the 'B' data (built using call to
+  /// Moose::FV::interpolateReconstruct). After construction, this object can be evaluated at both
+  /// element and face centers
+  CellCenteredMapFunctor<libMesh::VectorValue<ADReal>,
+                         std::unordered_map<dof_id_type, libMesh::VectorValue<ADReal>>>
+      _b2;
+
+  /// The subdomains that have body force residual objects
+  std::set<SubdomainID> _sub_ids_with_body_forces;
+
+  /// Whether we have performed our initial setup. Ordinarily we would do this in initialSetup but
+  /// there are wonky things that happen in other objects initialSetup that affect us, like how
+  /// Exodus output gathers all elements to 0 on distributed mesh
+  bool _initial_setup_done = false;
+
+private:
+  /**
+   * Pushes and pulls 'a' coefficient data between processes such that each process has all the 'a'
+   * coefficient data necessary to construct any requested Rhie-Chow velocity
+   */
+  void finalizeAData();
+
+  /**
+   * Performs the first and second overbar operations on 'B', e.g. one interpolation and one
+   * reconstruction
+   */
+  void computeFirstAndSecondOverBars();
+
+  /**
+   * Applies the interpolated and reconstructed 'B' data to the momentum residuals
+   */
+  void applyBData();
+
+  /**
+   * Pushes and pulls 'B' data between processes such that each process has all the 'B'
+   * data necessary to construct any requested Rhie-Chow velocity
+   */
+  void finalizeBData();
+
+  /// The velocity variable numbers
+  std::vector<unsigned int> _var_numbers;
+
+  /// Non-local elements that we should push and pull data for across processes
+  std::unordered_set<const Elem *> _elements_to_push_pull;
+
+  /// The nonlinear system
+  SystemBase & _sys;
+
+  /// An example datum used to help communicate AD vector information in parallel
+  const VectorValue<ADReal> _example;
+
+  /// If this is true, then interpolation and reconstruction operations will not be performed on the
+  /// body forces
+  const bool _standard_body_forces;
+
+  /// The x-component of the non-interpolated body forces
+  VectorComponentFunctor<ADReal> _bx;
+
+  /// The y-component of the non-interpolated body forces
+  VectorComponentFunctor<ADReal> _by;
+
+  /// The x-component of the interpolated and reconstructed body forces
+  VectorComponentFunctor<ADReal> _b2x;
+
+  /// The y-component of the interpolated and reconstructed body forces
+  VectorComponentFunctor<ADReal> _b2y;
```

Comment:
is this actually used? seems to me we work with _b ?
if so, where is _bz?

Reply:
It can be used for visualization purposes like in that input you commented on in order to help visualize the interpolated body force field. However, if you think that's dumb I can remove these. If you don't think it's dumb, I'll add the z-component

Reply:
I dont think it's dumb. And adding functors comes at no-cost anyway

Reply:
where are they filled btw? I only see them being declared

Reply:
They don't have to be filled. They just return a component of their parent functor

Reply:
nice. Z-component then? Or delete? Your call

Reply:
I added the z-component